File: leetcode_python/zigzag.py
# The string "PAYPALISHIRING" is written in a zigzag pattern on a given number of rows like this: (you may want to display this pattern in a fixed font for better legibility)

# P   A   H   N
# A P L S I I G
# Y   I   R
# And then read line by line: "PAHNAPLSIIGYIR"
# Write the code that will take a string and make this conversion given a number of rows:

# string convert(string text, int nRows);
# convert("PAYPALISHIRING", 3) should return "PAHNAPLSIIGYIR".

class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        # initialize a matrix with numRows * empty strings
        matrix = [""]*numRows
        # default moving direction, down
        inc = 1
        # i is the index of row in the matrix
        i = 0
        for x in range(len(s)):
            # append char in s to corresponding row of matrix
            matrix[i] += s[x]
            # if out of boundary, change moving direction
            if i+inc >=numRows or i+inc < 0:
                inc = -inc
            i += inc
        # append each strings in matrix to the result string     
        res = ""
        for ma in matrix:
            res += ma
        return res

File: leetcode_python/zigzag.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    var matrix = new Array(numRows).fill("");
    var inc = 1;
    var i = 0;
    Array.from(s).forEach((c)=>{
        matrix[i] += c;
        if(i+inc >= numRows && i-inc < 0) inc = 0;
        if(i+inc >= numRows || i+inc < 0){
            inc = -inc;
        }
        console.log(inc);
        i += inc;
    })
    
    var res = '';
    console.log(matrix);
    for(var m of matrix){
        res += m;
    }
    console.log(res);
    return res;
};

var s = 'AB';
var numRows = 1;
convert(s,numRows);